src/main/java/org/mozilla/jss/CryptoManager.c

```diff
@@ -517,6 +533,267 @@ Java_org_mozilla_jss_CryptoManager_initializeAllNative2
+        slotString == NULL ||
+        keySlotString == NULL ||
+        fipsString == NULL ||
+        fipsKeyString == NULL )
+    {
+        JSS_throw(env, NULL_POINTER_EXCEPTION);
+        goto finish;
+    }
+
+    /* Make sure initialize() completes only once */
+    if(initialized) {
+        JSS_throw(env, ALREADY_INITIALIZED_EXCEPTION);
+        goto finish;
+    }
+
+    /*
+     * Save the JavaVM pointer so we can retrieve the JNI environment
+     * later. This only works if there is only one Java VM.
+     */
+    if( (*env)->GetJavaVM(env, &JSS_javaVM) != 0 ) {
+        JSS_trace(env, JSS_TRACE_ERROR,
+                    "Unable to to access Java virtual machine");
+        PR_ASSERT(PR_FALSE);
+        goto finish;
+    }
+
+    /*
+     * Initialize the errcode translation table.
+     */
+    JSS_initErrcodeTranslationTable();
+
+    /*
+     * The rest of the initialization (the NSS stuff) is skipped if
+     * the initializeJavaOnly flag is set.
+     */
+    if( initializeJavaOnly) {
+        initialized = PR_TRUE;
+        goto finish;
+    }
+
+
+    /*
+     * Set the PKCS #11 strings
+     */
+    manuChars = JSS_RefJString(env, manuString);
+    libraryChars = JSS_RefJString(env, libraryString);
+    tokChars = JSS_RefJString(env, tokString);
+    keyTokChars = JSS_RefJString(env, keyTokString);
+    slotChars = JSS_RefJString(env, slotString);
+    keySlotChars = JSS_RefJString(env, keySlotString);
+    fipsChars = JSS_RefJString(env, fipsString);
+    fipsKeyChars = JSS_RefJString(env, fipsKeyString);
+    if( (*env)->ExceptionOccurred(env) ) {
+        ASSERT_OUTOFMEM(env);
+        goto finish;
+    }
+    PR_ASSERT( strlen(manuChars) == 33 );
+    PR_ASSERT( strlen(libraryChars) == 33 );
+    PR_ASSERT( strlen(tokChars) == 33 );
+    PR_ASSERT( strlen(keyTokChars) == 33 );
+    PR_ASSERT( strlen(slotChars) == 65 );
+    PR_ASSERT( strlen(keySlotChars) == 65 );
+    PR_ASSERT( strlen(fipsChars) == 65 );
+    PR_ASSERT( strlen(fipsKeyChars) == 65 );
+
+    NSSInitParameters initparams;
+    memset( &initparams, 0, sizeof( initparams ) );
+    initparams.length = sizeof( initparams );
+
+    initparams.minPWLen = 0;
+    initparams.manufactureID = (char *) manuChars;
+    initparams.libraryDescription = (char *) libraryChars;
+    initparams.cryptoTokenDescription = (char *) tokChars;
+    initparams.dbTokenDescription = (char *) keyTokChars;
+    initparams.FIPSTokenDescription = (char *) fipsChars;
+    initparams.cryptoSlotDescription =(char *) slotChars;
+    initparams.dbSlotDescription = (char *) keySlotChars;
+    initparams.FIPSSlotDescription = (char *) fipsKeyChars;
+    initparams.passwordRequired = PR_FALSE;
+	
+    szConfigDir = JSS_RefJString(env, configDir);
+    /*
+    * Set up arguments to NSS_InitContext
+    */
+    szCertPrefix = JSS_RefJString(env, certPrefix);
+    szKeyPrefix = JSS_RefJString(env, keyPrefix);
+    szSecmodName = JSS_RefJString(env, secmodName);
+
```

Comment:
Shouldn't the above three szXXX param assignments still need to be under
"    if( certPrefix != NULL || keyPrefix != NULL || secmodName != NULL "
like the original code?

Reply:
This ok for the following reason:

The routine JSS_RefJString simply returns NULL if the input is NULL. That check in the original function was done to see if we can use one of the simpler NSS_Init variants. In this case we have only one NSS_InitContext so we simply will pass NULL for those 3 values of interest.